connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/utils/SchemaConverters.scala

```diff
@@ -92,14 +109,38 @@ object SchemaConverters {
             MapType(keyType, valueType, valueContainsNull = false).defaultConcreteType,
             nullable = false))
       case MESSAGE =>
-        if (existingRecordNames.contains(fd.getFullName)) {
-          throw QueryCompilationErrors.foundRecursionInProtobufSchema(fd.toString())
+        // User can set circularReferenceDepth of 0 or 1 or 2.
+        // Going beyond 3 levels of recursion is not allowed.
+        if (protobufOptions.circularReferenceType.equals("FIELD_TYPE")) {
+          if (existingRecordTypes.contains(fd.getType.name()) &&
+            (protobufOptions.circularReferenceDepth < 0 ||
+              protobufOptions.circularReferenceDepth >= 3)) {
+            throw QueryCompilationErrors.foundRecursionInProtobufSchema(fd.toString())
+          } else if (existingRecordTypes.contains(fd.getType.name()) &&
```

Comment:
name or full name? 
also what keeps track of the recursion depth?

Reply:
@rangadi we have two maps with incremental counters, one for field_name base check and one for field_type.

Reply:
In my back-ported branch,
```
        val recordName = circularReferenceType match {
          case CircularReferenceTypes.FIELD_NAME =>
            fd.getFullName
          case CircularReferenceTypes.FIELD_TYPE =>
            fd.getFullName().substring(0, fd.getFullName().lastIndexOf(".")) 
        }
        
        if (circularReferenceTolerance < 0 && existingRecordNames(recordName) > 0) {
          // no tolerance on circular reference
          logError(s"circular reference in protobuf schema detected [no tolerance] - ${recordName}")
          throw new IllegalStateException(s"circular reference in protobuf schema detected [no tolerance] - ${recordName}")
        }

        if (existingRecordNames(recordName) > (circularReferenceTolerance max 0) ) {
          // stop navigation and drop the repetitive field
          logInfo(s"circular reference in protobuf schema detected [max tolerance breached] field dropped - ${recordName} = ${existingRecordNames(recordName)}")
          Some(NullType)
        } else {
          val newRecordNames: Map[String, Int] = existingRecordNames +  
            (recordName -> (1 + existingRecordNames(recordName)))
          Option(
            fd.getMessageType.getFields.asScala
              .flatMap(structFieldFor(_, newRecordNames, protobufOptions))
              .toSeq)
            .filter(_.nonEmpty)
            .map(StructType.apply)
        }```